src/Database/QueryCompiler.php

```
@@ -102,11 +102,13 @@ public function compile(Query $query, ValueBinder $binder): string
         );
 
         // Propagate bound parameters from sub-queries if the
-        // placeholders can be found in the SQL statement.
+        // placeholders can be found in the SQL statement. Only
+        // add new placeholders, as sub-queries may have been executed already.
         if ($query->getValueBinder() !== $binder) {
+            $existing = $binder->bindings();
             foreach ($query->getValueBinder()->bindings() as $binding) {
                 $placeholder = ':' . $binding['placeholder'];
-                if (preg_match('/' . $placeholder . '(?:\W|$)/', $sql) > 0) {
+                if (!isset($existing[$placeholder]) && preg_match('/' . $placeholder . '(?:\W|$)/', $sql) > 0) {
```

Did we not have tests that generated invalid duplicate placeholders?

---

We didn't. We do now though.

---

I'm just trying to think if there's any need to throw an error when this exists outside of the issue you're fixing.

Otherwise, looks good to me.